File: LaTeXDraw/src/net/sf/latexdraw/actions/ScaleShapes.java
package net.sf.latexdraw.actions;

import java.awt.geom.Rectangle2D;

import net.sf.latexdraw.glib.models.interfaces.GLibUtilities;
import net.sf.latexdraw.glib.models.interfaces.IGroup;
import net.sf.latexdraw.glib.models.interfaces.IPoint;
import net.sf.latexdraw.glib.models.interfaces.IShape.Position;

import org.malai.undo.Undoable;

/**
 * This action scales the selected shapes of a drawing.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2012 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 29/11/2011<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
public class ScaleShapes extends ShapeAction<IGroup> implements Undoable, Modifying {
	/** The direction of the scaling. */
	protected Position refPosition;

	/** The new X position used to compute the scale factor. */
	protected double newX;

	/** The new Y position used to compute the scale factor. */
	protected double newY;

	/** The bound of the selected shapes used to perform the scaling. */
	private Rectangle2D bound;

	/** The old width of the selection. */
	private double oldWidth;

	/** The old height of the selection. */
	private double oldHeight;


	/**
	 * Creates the action.
	 */
	public ScaleShapes() {
		super();
		oldWidth	= Double.NaN;
		oldHeight	= Double.NaN;
		newX		= Double.NaN;
		newY		= Double.NaN;
		bound 		= new Rectangle2D.Double();
	}


	@Override
	public boolean isRegisterable() {
		return hadEffect();
	}


	@Override
	public boolean canDo() {
		return super.canDo() && refPosition!=null && isValidScales();
	}


	private boolean isValidScales() {
		switch(refPosition) {
			case EAST: case WEST:
				return isValidScale(newX) && getScaledWidth()>1.;
			case NORTH: case SOUTH:
				return isValidScale(newY) && getScaledHeight()>1.;
			default:
				return isValidScale(newX) && isValidScale(newY) && getScaledHeight()>1. && getScaledWidth()>1.;
		}
	}


	private boolean isValidScale(final double scale) {
		return GLibUtilities.INSTANCE.isValidCoordinate(scale) && scale>0;
	}


	@Override
	protected void doActionBody() {
		if(Double.isNaN(oldWidth)) {
			final IPoint br = shape.getBottomRightPoint();
			final IPoint tl = shape.getTopLeftPoint();
			oldWidth  = br.getX() - tl.getX();
			oldHeight = br.getY() - tl.getY();
			updateBound(tl, br);
		}
		redo();
	}


	private void updateBound(final IPoint tl, final IPoint br) {
		bound.setFrameFromDiagonal(tl.getX(), tl.getY(), br.getX(), br.getY());
	}


	@Override
	public void undo() {
		shape.scale(oldWidth, oldHeight, refPosition, bound);
		drawing.setModified(true);
		updateBound(shape.getTopLeftPoint(), shape.getBottomRightPoint());
	}


	@Override
	public void setShape(final IGroup shape) {
		super.setShape(shape);

		if(shape!=null)
			updateBound(shape.getTopLeftPoint(), shape.getBottomRightPoint());
	}


	private double getScaledHeight() {
		if(refPosition.isSouth())
			return bound.getHeight() + bound.getY() - newY;
		else if(refPosition.isNorth())
			return bound.getHeight() + newY - bound.getMaxY();
		return 0.;
	}


	private double getScaledWidth() {
		if(refPosition.isWest())
			return bound.getWidth() + newX - bound.getMaxX();
		else if(refPosition.isEast())
			return bound.getWidth() + bound.getX() - newX;
		return 0.;
	}


	@Override
	public void redo() {
		shape.scale(getScaledWidth(), getScaledHeight(), refPosition, bound);
		drawing.setModified(true);
		updateBound(shape.getTopLeftPoint(), shape.getBottomRightPoint());
	}


	@Override
	public String getUndoName() {
		return "Resizing";
	}


	/**
	 * @return The reference position of the scaling.
	 * @since 3.0
	 */
	public Position getRefPosition() {
		return refPosition;
	}


	/**
	 * @param newX The new X position used to compute the scale factor.
	 * @since 3.0
	 */
	public final void setNewX(final double newX) {
		this.newX = newX;
	}


	/**
	 * @param newY The new Y position used to compute the scale factor.
	 * @since 3.0
	 */
	public final void setNewY(final double newY) {
		this.newY = newY;
	}


	/**
	 * @param refPosition The reference position of the scaling.
	 * @since 3.0
	 */
	public void setRefPosition(final Position refPosition) {
		this.refPosition = refPosition;
	}
}

File: latexdraw-research/GUImutants/mutant23/net.sf.latexdraw.mutant23/src/main/net/sf/latexdraw/actions/shape/ScaleShapes.scala
package net.sf.latexdraw.actions.shape

import java.awt.geom.Rectangle2D
import org.malai.action.Action
import org.malai.undo.Undoable
import net.sf.latexdraw.actions.DrawingAction
import net.sf.latexdraw.actions.Modifying
import net.sf.latexdraw.actions.ShapeAction
import net.sf.latexdraw.glib.models.interfaces.IShape.Position
import net.sf.latexdraw.glib.models.interfaces.GLibUtilities
import net.sf.latexdraw.glib.models.interfaces.IGroup
import net.sf.latexdraw.glib.models.interfaces.IPoint
import net.sf.latexdraw.util.LNumber

/**
 * This action scales a shape.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2013 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 2013-03-07<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
class ScaleShapes extends Action with ShapeAction[IGroup] with DrawingAction with Undoable with Modifying {
	/** The direction of the scaling. */
	protected var _refPosition : Option[Position] = None

	/** The new X position used to compute the scale factor. */
	protected var _newX : Double = Double.NaN

	/** The new Y position used to compute the scale factor. */
	protected var _newY : Double = Double.NaN

	/** The bound of the selected shapes used to perform the scaling. */
	private val bound : Rectangle2D = new Rectangle2D.Double()

	/** The old width of the selection. */
	private var oldWidth : Double = Double.NaN

	/** The old height of the selection. */
	private var oldHeight : Double = Double.NaN

	private var doneOnce = false



	override def isRegisterable() = hadEffect

	override def hadEffect = isDone && (!LNumber.INSTANCE.equals(oldWidth, bound.getWidth) || !LNumber.INSTANCE.equals(oldHeight, bound.getHeight))

	override def canDo() = _drawing.isDefined && _shape.isDefined && _refPosition.isDefined && isValidScales


	private def isValidScales = {
		_refPosition.get match {
			case Position.EAST => isValidScale(_newX) && scaledWidth(_newX)>1.0
			case Position.WEST => isValidScale(_newX) && scaledWidth(_newX)>1.0
			case Position.NORTH => isValidScale(_newY) && scaledHeight(_newY)>1.0
			case Position.SOUTH => isValidScale(_newY) && scaledHeight(_newY)>1.0
			case _ => isValidScale(_newX) && isValidScale(_newY) && scaledHeight(_newY)>1.0 && scaledWidth(_newX)>1.0
		}
	}


	private def isValidScale(scale : Double) = GLibUtilities.INSTANCE.isValidCoordinate(scale) && scale>0


	protected def doActionBody() {
		if(oldWidth.isNaN) {
			val br = _shape.get.getBottomRightPoint
			val tl = _shape.get.getTopLeftPoint
			oldWidth  = br.getX - tl.getX
			oldHeight = br.getY - tl.getY
			updateBound(tl, br)
		}
		redo
	}


	private def updateBound(tl : IPoint, br : IPoint) {
		bound.setFrameFromDiagonal(tl.getX, tl.getY, br.getX, br.getY)
	}


	override def undo() {
		val sh = _shape.get
		//Mutant23
		//sh.scale(oldWidth, oldHeight, _refPosition.get, bound)
		sh.setModified(true)
		_drawing.get.setModified(true)
		updateBound(sh.getTopLeftPoint, sh.getBottomRightPoint)
	}


	override def setShape(shape : IGroup) {
		super.setShape(shape)

		if(shape!=null)
			updateBound(shape.getTopLeftPoint, shape.getBottomRightPoint)
	}


	private def scaledHeight(y:Double) : Double = {
		_refPosition.get match {
			case ref if(ref.isSouth) => bound.getHeight + bound.getY - y
			case ref if(ref.isNorth) =>  bound.getHeight + y - bound.getMaxY
			case _ => 0.0
		}
	}


	private def scaledWidth(x:Double) : Double = {
		_refPosition.get match {
			case ref if(ref.isWest) => bound.getWidth + x - bound.getMaxX
			case ref if(ref.isEast) => bound.getWidth + bound.getX - x
			case _ => 0.0
		}
	}


	override def redo() {
		val sh = _shape.get
		sh.scale(scaledWidth(_newX), scaledHeight(_newY), _refPosition.get, bound)
		sh.setModified(true)
		_drawing.get.setModified(true)
		updateBound(sh.getTopLeftPoint, sh.getBottomRightPoint)
	}


	override def getUndoName() = "Resizing"


	def refPosition = _refPosition

	/**
	 * @param refPosition The reference position of the scaling.
	 */
	def refPosition_=(refPosition : Position) {
		_refPosition = Some(refPosition)
	}

	def newX = _newX

	/**
	 * @param newX The new X position used to compute the scale factor.
	 */
	def newX_=(newX : Double) {
		if(scaledWidth(newX)>1)
			_newX = newX
	}

	def newY = _newY

	/**
	 * @param newY The new Y position used to compute the scale factor.
	 */
	def newY_=(newY : Double) {
		if(scaledHeight(newY)>1)
			_newY = newY
	}
}
